phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1642,7 +1704,9 @@ private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put at
                   transferAttributes(atomicPut, put);
                   mutations.add(put);
               }
-              put.add(cell);
+              if (checkCellNeedUpdate(cell, currCfCqCellExprMap)) {
```

Comment:
I think if you do this check before allocating the put object and adding it to the mutations list it will simplify your code.

Reply:
@tkhurana I don't get your point. Adding the put object to mutations only need to be done once, but adding cell to put should be done each time when the cell need to be updated. What is the difference if we do this check earlier?